datalad/core/dataset/dataset.py

```diff
@@ -0,0 +1,267 @@
+        if not op.isabs(path):
+            path = op.join(op.getpwd(), path)
+
+        # use canonical paths only:
+        return op.normpath(path)
+
+    def _flyweight_invalid(self):
+        """Invalidation of Flyweight instance
+
+        Dataset doesn't need to be invalidated during its lifetime at all. Instead the underlying *Repo instances are.
+        Dataset itself can represent a not yet existing path.
+        """
+        return False
+    # End Flyweight
+
+    def __hash__(self):
+        # the flyweight key is already determining unique instances
+        # add the class name to distinguish from strings of a path
+        return hash((self.__class__.__name__, self.__weakref__.key))
+
+    def __init__(self, path):
+        """
+        Parameters
+        ----------
+        path : str or Path
+          Path to the dataset location. This location may or may not exist
+          yet.
+        """
+        self._pathobj = path if isinstance(path, Path) else None
+        if isinstance(path, PurePath):
+            path = str(path)
+        self._path = path
+        self._repo = None
+        self._id = None
+        self._cfg = None
+        self._cfg_bound = None
+
+    @property
+    def pathobj(self):
+        """pathobj for the dataset"""
+        # XXX this relies on the assumption that self._path as managed
+        # by the base class is always a native path
+        if not self._pathobj:
+            self._pathobj = Path(self._path)
+        return self._pathobj
+
+    def __eq__(self, other):
+        if not hasattr(other, 'pathobj'):
+            return False
+        # Ben: https://github.com/datalad/datalad/pull/4057#discussion_r370153586
+        # It's pointing to the same thing, while not being the same object
+        # (in opposition to the *Repo classes). So `ds1 == ds2`,
+        # `but ds1 is not ds2.` I thought that's a useful distinction. On the
+        # other hand, I don't think we use it anywhere outside tests yet.
+        me_exists = self.pathobj.exists()
+        other_exists = other.pathobj.exists()
+        if me_exists != other_exists:
+            # no chance this could be the same
+            return False
+        elif me_exists:
+            # check on filesystem
+            return self.pathobj.samefile(other.pathobj)
+        else:
+            # we can only do lexical comparison.
+            # this will fail to compare a long and a shortpath.
+            # on windows that could actually point to the same thing
+            # if it would exists, but this is how far we go with this.
+            return self.pathobj == other.pathobj
+
+    def close(self):
+        """Perform operations which would close any possible process using this Dataset
+        """
+        repo = self._repo
+        self._repo = None
+        if repo:
+            # might take care about lingering batched processes etc
+            del repo
+
+    @property
```

Comment:
I wonder whether we want to put `path` into core. We prob. should still aim to switch to `pathlib` entirely. If `path` remains in core, we would need to break it later or not get rid of `os.path`. How about renaming to `_std_path` or something like that as a compromise to allow for this move w/o the need to do all the pathlib related work first while making it (at least conventionally) clear, that this is nothing an user should rely on?

Reply:
I think you are right, I will try to remove it.

Reply:
Done